crates/ra_hir_ty/src/utils.rs

```diff
@@ -62,6 +62,27 @@ pub(super) fn all_super_traits(db: &impl DefDatabase, trait_: TraitId) -> Vec<Tr
     result
 }
 
+/// Finds a path from a trait to one of its descendant traits. Returns an empty
+/// vector if there is no path.
+pub(super) fn find_super_trait_path(
+    db: &impl DefDatabase,
+    super_trait: TraitId,
+    trait_: TraitId,
+) -> Vec<TraitId> {
+    if trait_ == super_trait {
+        return vec![trait_];
+    }
+
+    for tt in direct_super_traits(db, trait_) {
+        let mut path = find_super_trait_path(db, super_trait, tt);
```

Comment:
Should we guard against cycles here?

Reply:
Should we maybe re-use `all_super_traits`? Oh, we can't, because that flattens the path

Reply:
Yeah, I was vaguely thinking that we can't run into cycles because we used `all_super_traits` before, but that's of course not true...